### PR TITLE
[KYUUBI #886] Add HTTP transport mode support to KYUUBI - no Kerberos support

### PR DESCRIPTION
**Pull request by mahmoudbahaa**
Same as https://github.com/apache/incubator-kyuubi/pull/2815 but with Kerberos auth removed as was not complete nor tested. so thought to make a separated PR with no Kerberos support then we can add Kerberos support later on

### _Why are the changes needed?_

Add http transport protocol support to Kyuubi
main code was taken from:

1. https://github.com/SteNicholas/incubator-kyuubi/commit/25ac0731d7331616988026996ca918793af151e9
2. https://github.com/apache/hive/blob/branch-3.1/service/src/java/org/apache/hive/service/cli/thrift/ThriftHttpCLIService.java
2. https://github.com/apache/hive/blob/branch-3.1/service/src/java/org/apache/hive/service/cli/thrift/ThriftHttpServlet.java

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
we can remove the current tests in this PR and add a test like this, which covers both client/server behavior.

```scala
package org.apache.kyuubi.operation.thift.http

class KyuubiOperationPerUserSuite extends WithKyuubiServer with SparkQueryTests {

  override protected val conf: KyuubiConf = KyuubiConf()

  override protected val frontendProtocols: Seq[KyuubiConf.FrontendProtocols.Value] =
    FrontendProtocols.THRIFT_HTTP :: Nil

  override protected def getJdbcUrl: String =
    s"jdbc:hive://${server.frontendServices.head.connectionUrl}/;transportMode=http"

  override protected def jdbcUrl: String = getJdbcUrl
}
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/util/HttpAuthUtils.scala [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+import org.apache.kyuubi.Logging
+
+object HttpAuthUtils {
+  val WWW_AUTHENTICATE = "WWW-Authenticate"
+  val AUTHORIZATION = "Authorization"
+  val BASIC = "Basic"
+  val NEGOTIATE = "Negotiate"
+
+  private val logger = new Logger
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/util/HttpAuthUtils.scala:34]:
```suggestion
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/util/HttpAuthUtils.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  /**
+   * Parses a cookie token to retrieve client user name.
+   *
+   * @param tokenStr Token String.
+   * @return A valid user name if input is of valid format, else returns null.
+   */
+  def getUserNameFromCookieToken(tokenStr: String): String = {
+    val map = splitCookieToken(tokenStr)
+    if (!(map.keySet == COOKIE_ATTRIBUTES)) {
+      logger.error("Invalid token with missing attributes " + tokenStr)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/util/HttpAuthUtils.scala:67]:
```suggestion
      error("Invalid token with missing attributes " + tokenStr)
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/util/HttpAuthUtils.scala [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        return null
+      }
+      val key = part.substring(0, separator)
+      val value = part.substring(separator + 1)
+      map.put(key, value)
+    }
+    map
+  }
+
+  private class Logger extends Logging
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/util/HttpAuthUtils.scala:99]:
```suggestion
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/util/SessionManager.scala [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+  def getProxyUserName: String = {
+    return threadLocalProxyUserName.get
+  }
+
+  def clearProxyUserName(): Unit = {
+    threadLocalProxyUserName.remove()
+  }
+
+  private class Logger extends Logging
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/util/SessionManager.scala:92]:
```suggestion
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+  }
+
+  override protected def getIpAddress: String = {
+    SessionManager.getIpAddress
+  }
+
+  override protected def getUserName(req: TOpenSessionReq): String = {
+    var userName: String = SessionManager.getUserName
+    if (userName == null) userName = req.getUsername
+    userName = getShortName(userName)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:267]: we can use org.apache.kyuubi.service.ServiceUtils#getShortName

**Review by yaooqinn — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merging to master

**Pull request closed without merging**